sdks/java/io/hadoop-input-format/src/main/java/org/apache/beam/sdk/io/hadoop/inputformat/HadoopInputFormatIO.java

```diff
@@ -0,0 +1,1121 @@
+           * are {@link org.apache.hadoop.mapreduce.lib.db.DBInputFormat DBInputFormat}, {@link
+           * org.apache.hadoop.hbase.mapreduce.TableInputFormat TableInputFormat}, etc.
+           */
+          if (Configurable.class.isAssignableFrom(inputFormatObj.getClass())) {
+            ((Configurable) inputFormatObj).setConf(conf.getHadoopConfiguration());
+          }
+        } catch (InstantiationException e) {
+          throw new IOException("Unable to create InputFormat object: ", e);
+        } catch (IllegalAccessException e) {
+          throw new IOException("Unable to create InputFormat object: ", e);
+        } catch (ClassNotFoundException e) {
+          throw new IOException("Unable to create InputFormat object: ", e);
+        }
+      }
+    }
+
+    /**
+     * Throws exception if you set different InputFormat key or value class than InputFormat's
+     * actual key or value class. If you set incorrect classes then, it may result in an error like
+     * "unexpected extra bytes after decoding" while the decoding process happens. Hence this
+     * validation is required.
+     */
+    private void validateUserInputForKeyAndValue() throws IOException, InterruptedException {
+      ParameterizedType genericClassType = determineGenericType();
+      boolean isCorrectKeyClassSet = validateClasses(
+          genericClassType.getActualTypeArguments()[0].getTypeName(),
+          keyCoder,
+          "key.class");
+      boolean isCorrectValueClassSet = validateClasses(
+          genericClassType.getActualTypeArguments()[1].getTypeName(),
+          valueCoder,
+          "value.class");
+      if (!isCorrectKeyClassSet) {
+        Class<?> actualClass = conf.getHadoopConfiguration().getClass("key.class", Object.class);
+        throw new IllegalArgumentException(String.format(
+            HadoopInputFormatIOConstants.WRONG_INPUTFORMAT_KEY_CLASS_ERROR_MSG,
+            getExpectedKeyClass().getName(),
+            actualClass.getName()));
+      }
+      if (!isCorrectValueClassSet) {
+        Class<?> actualClass = conf.getHadoopConfiguration().getClass("value.class", Object.class);
+        throw new IllegalArgumentException(String.format(
+            HadoopInputFormatIOConstants.WRONG_INPUTFORMAT_VALUE_CLASS_ERROR_MSG,
+            getExpectedValueClass().getName(),
+            actualClass.getName()));
+      }
+    }
+
+    /**
+     * Returns true if key/value class set by the user is compatible with the key/value class of a
+     * pair returned by RecordReader. User provided key and value classes are validated against
+     * parameterized type of InputFormat. If parameterized type has any type parameter such as T,
+     * K, V, etc then validation is done by encoding and decoding first pair returned by
+     * RecordReader.
+     */
+    private <T> boolean validateClasses(
+        String inputFormatGenericType,
+        Coder<T> coder,
+        String property) throws IOException, InterruptedException {
```

Comment:
for property parameter:
in both places where this is used, it seems like the actual thing you want is an instance of the class type & the key and/or value itself. Why not just pass those in rather than pass around the property value? that lets you remove a lot of if statements.